elements/lisk-codec/src/collection.ts

```diff
@@ -177,76 +175,57 @@ export const readArray = (message: Buffer, offset: number, compiledSchema: Compi
 		return [[], index];
 	}
 	const startingByte = message[index];
-	// It should be wire type 2 as it's object
-	const [, keySize] = readUInt32(message, index);
-	index += keySize;
-	// Takeout the length
-	const [arrayLength, wireType2Size] = readUInt32(message, index);
-	index += wireType2Size;
-	const [, typeSchema] = compiledSchema;
+	const [rootSchema, typeSchema] = compiledSchema;
+	// Peek the current key, and if not the same fieldnumber, skip
+	const [key] = readUInt32(message, index);
+	const [fieldNumber] = readKey(key);
+	if (fieldNumber !== (rootSchema as CompiledSchema).schemaProp.fieldNumber) {
+		return [[], index];
+	}
 	// eslint-disable-next-line @typescript-eslint/no-explicit-any
 	const result: Array<any> = [];
 	// Case of object
 	if (Array.isArray(typeSchema)) {
 		// handle as object
 		const [nestedTypeSchema] = typeSchema;
 		if (nestedTypeSchema.schemaProp.type === 'object') {
-			// Since first key and length is read, we do read them directly
-			if (wireType2Size !== 0) {
-				// readObject returns Next offset, not index used
-				const [objectResult, nextOffset] = readObject(message, index, typeSchema);
-				result.push(objectResult);
-				index = nextOffset;
-			} else {
-				result.push({});
-			}
 			// If still the next bytes is the same key, it is still element of array
 			while (message[index] === startingByte) {
 				const [, wire2KeySize] = readUInt32(message, index);
 				index += wire2KeySize;
 				// Takeout the length
 				const [wireType2Length, wireType2LengthSize] = readUInt32(message, index);
+				// for object, length is not used
 				index += wireType2LengthSize;
 				if (wireType2Length === 0) {
-					result.push({});
 					// Add default value
+					result.push({});
 					continue;
 				}
 				// readObject returns Next offset, not index used
 				const [res, nextOffset] = readObject(message, index, typeSchema);
 				result.push(res);
-				index += nextOffset;
+				index = nextOffset;
```

Comment:
Not sure I completely understand but I hope this is an intentional change. `index` will always be `nextOffset`

Reply:
This is intentional.
`readObject` and `readArray` return the nextOffset, but the scalar value returns the byte size it used.
Therefore it's correct to assign to the `index`